File: packages/python/src/qagentic/core/reporter.py
"""
Reporter classes for outputting test results to various destinations.
"""

import json
import os
from abc import ABC, abstractmethod
from datetime import datetime, timezone
from pathlib import Path
from typing import Any, Dict, List, Optional, TYPE_CHECKING
import threading

from qagentic.core.config import get_config, QAgenticConfig
from qagentic.core.test_result import TestResult, TestRunResult
from qagentic.core.status import Status

if TYPE_CHECKING:
    pass


class BaseReporter(ABC):
    """Base class for all reporters."""
    
    @abstractmethod
    def start_run(self, run: TestRunResult) -> None:
        """Called when a test run starts."""
        pass
    
    @abstractmethod
    def end_run(self, run: TestRunResult) -> None:
        """Called when a test run ends."""
        pass
    
    @abstractmethod
    def report_test(self, test: TestResult) -> None:
        """Report a single test result."""
        pass


class ConsoleReporter(BaseReporter):
    """Rich console output reporter."""
    
    def __init__(self, config: Optional[QAgenticConfig] = None) -> None:
        self.config = config or get_config()
        self._use_rich = self.config.features.rich_console
        
        if self._use_rich:
            try:
                from rich.console import Console
                from rich.table import Table
                from rich.panel import Panel
                self._console = Console()
                self._rich_available = True
            except ImportError:
                self._rich_available = False
                self._use_rich = False
        else:
            self._rich_available = False
    
    def start_run(self, run: TestRunResult) -> None:
        """Print run start banner."""
        if self._use_rich and self._rich_available:
            from rich.panel import Panel
            self._console.print(Panel(
                f"[bold blue]QAagentic Test Run[/bold blue]\n"
                f"Project: {run.project_name}\n"
                f"Environment: {run.environment}",
                title="🚀 Starting Tests",
                border_style="blue",
            ))
        else:
            print(f"\n{'='*60}")
            print(f"QAagentic Test Run - {run.project_name}")
            print(f"Environment: {run.environment}")
            print(f"{'='*60}\n")
    
    def end_run(self, run: TestRunResult) -> None:
        """Print run summary."""
        if self._use_rich and self._rich_available:
            from rich.table import Table
            from rich.panel import Panel
            
            table = Table(show_header=True, header_style="bold")
            table.add_column("Status", style="bold")
            table.add_column("Count", justify="right")
            
            table.add_row("[green]Passed[/green]", str(run.passed))
            table.add_row("[red]Failed[/red]", str(run.failed))
            table.add_row("[yellow]Broken[/yellow]", str(run.broken))
            table.add_row("[blue]Skipped[/blue]", str(run.skipped))
            table.add_row("[bold]Total[/bold]", str(run.total))
            
            status_color = "green" if run.is_successful else "red"
            status_icon = "✅" if run.is_successful else "❌"
            
            self._console.print(Panel(
                table,
                title=f"{status_icon} Test Run Complete - {run.pass_rate:.1f}% Pass Rate",
                border_style=status_color,
            ))
        else:
            print(f"\n{'='*60}")
            print(f"Test Run Complete - {run.pass_rate:.1f}% Pass Rate")
            print(f"Passed: {run.passed} | Failed: {run.failed} | Skipped: {run.skipped}")
            print(f"{'='*60}\n")
    
    def report_test(self, test: TestResult) -> None:
        """Print test result."""
        if not self.config.features.console_output:
            return
        
        status_symbols = {
            Status.PASSED: ("✓", "green"),
            Status.FAILED: ("✗", "red"),
            Status.BROKEN: ("!", "yellow"),
            Status.SKIPPED: ("○", "blue"),
        }
        
        symbol, color = status_symbols.get(test.status, ("?", "white"))
        
        if self._use_rich and self._rich_available:
            self._console.print(f"  [{color}]{symbol}[/{color}] {test.name}")
            if test.error_message:
                self._console.print(f"    [dim red]{test.error_message[:100]}...[/dim red]")
        else:
            print(f"  {symbol} {test.name}")
            if test.error_message:
                print(f"    Error: {test.error_message[:100]}...")


class JSONReporter(BaseReporter):
    """JSON file reporter."""
    
    def __init__(self, config: Optional[QAgenticConfig] = None) -> None:
        self.config = config or get_config()
        self.output_dir = Path(self.config.local.output_dir)
        self._tests: List[dict] = []
        self._run: Optional[TestRunResult] = None
    
    def start_run(self, run: TestRunResult) -> None:
        """Initialize for new run."""
        self._run = run
        self._tests = []
        
        # Create output directory
        self.output_dir.mkdir(parents=True, exist_ok=True)
        
        # Clean previous results if configured
        if self.config.local.clean_on_start:
            for f in self.output_dir.glob("*.json"):
                f.unlink()
    
    def end_run(self, run: TestRunResult) -> None:
        """Write final JSON report."""
        if not self.config.local.enabled or "json" not in self.config.local.formats:
            return
        
        # Write run summary
        run_file = self.output_dir / "run.json"
        with open(run_file, "w") as f:
            json.dump(run.to_dict(), f, indent=2, default=str)
        
        # Write individual test results
        tests_dir = self.output_dir / "tests"
        tests_dir.mkdir(exist_ok=True)
        
        for test in run.tests:
            test_file = tests_dir / f"{test.id}.json"
            with open(test_file, "w") as f:
                json.dump(test.to_dict(), f, indent=2, default=str)
    
    def report_test(self, test: TestResult) -> None:
        """Add test to collection."""
        self._tests.append(test.to_dict())


class JUnitReporter(BaseReporter):
    """JUnit XML reporter for CI/CD compatibility."""
    
    def __init__(self, config: Optional[QAgenticConfig] = None) -> None:
        self.config = config or get_config()
        self.output_dir = Path(self.config.local.output_dir)
        self._run: Optional[TestRunResult] = None
    
    def start_run(self, run: TestRunResult) -> None:
        """Initialize for new run."""
        self._run = run
        self.output_dir.mkdir(parents=True, exist_ok=True)
    
    def end_run(self, run: TestRunResult) -> None:
        """Write JUnit XML report."""
        if not self.config.local.enabled or "junit" not in self.config.local.formats:
            return
        
        import xml.etree.ElementTree as ET
        
        # Create testsuite element
        testsuite = ET.Element("testsuite")
        testsuite.set("name", run.project_name)
        testsuite.set("tests", str(run.total))
        testsuite.set("failures", str(run.failed))
        testsuite.set("errors", str(run.broken))
        testsuite.set("skipped", str(run.skipped))
        testsuite.set("time", str(run.duration_ms / 1000))
        testsuite.set("timestamp", run.start_time.isoformat() if run.start_time else "")
        
        for test in run.tests:
            testcase = ET.SubElement(testsuite, "testcase")
            testcase.set("name", test.name)
            testcase.set("classname", test.class_name or test.module or "")
            testcase.set("time", str(test.duration_ms / 1000))
            
            if test.status == Status.FAILED:
                failure = ET.SubElement(testcase, "failure")
                failure.set("message", test.error_message or "Test failed")
                failure.set("type", test.error_type or "AssertionError")
                if test.stack_trace:
                    failure.text = test.stack_trace
            
            elif test.status == Status.BROKEN:
                error = ET.SubElement(testcase, "error")
                error.set("message", test.error_message or "Test error")
                error.set("type", test.error_type or "Error")
                if test.stack_trace:
                    error.text = test.stack_trace
            
            elif test.status == Status.SKIPPED:
                skipped = ET.SubElement(testcase, "skipped")
                if test.error_message:
                    skipped.set("message", test.error_message)
        
        # Write to file
        tree = ET.ElementTree(testsuite)
        junit_file = self.output_dir / "junit.xml"
        tree.write(junit_file, encoding="unicode", xml_declaration=True)
    
    def report_test(self, test: TestResult) -> None:
        """No-op for JUnit - all written at end."""
        pass


class APIReporter(BaseReporter):
    """Reporter that sends results to QAagentic API."""
    
    def __init__(self, config: Optional[QAgenticConfig] = None) -> None:
        self.config = config or get_config()
        self._client: Optional[Any] = None
        self._run: Optional[TestRunResult] = None
        self._batch: List[TestResult] = []
    
    def _get_client(self) -> Any:
        """Get or create HTTP client."""
        if self._client is None:
            import httpx
            self._client = httpx.Client(
                base_url=self.config.api.url,
                timeout=self.config.api.timeout,
                headers={
                    "Content-Type": "application/json",
                    "X-API-Key": self.config.api.key or "",
                    "X-Project": self.config.project_name,
                },
            )
        return self._client
    
    def start_run(self, run: TestRunResult) -> None:
        """Register run with API."""
        if not self.config.api.enabled:
            return
        
        self._run = run
        self._batch = []
        
        try:
            client = self._get_client()
            # Use correct API Gateway endpoint with camelCase fields
            response = client.post("/api/test-runs", json={
                "projectName": run.project_name,
                "environment": run.environment,
                "startTime": run.start_time.isoformat() if run.start_time else None,
                "metadata": {
                    "pythonVersion": None,
                    "platform": None,
                    "testFramework": "pytest"
                }
            })
            response.raise_for_status()
            data = response.json()
            # Store the API-generated run ID
            self._api_run_id = data.get('runId', run.id)
        except Exception as e:
            # Log but don't fail tests
            print(f"Warning: Failed to register run with API: {e}")
            self._api_run_id = run.id
    
    def end_run(self, run: TestRunResult) -> None:
        """Finalize run with API."""
        if not self.config.api.enabled:
            return
        
        # Flush any remaining batch
        self._flush_batch()
        
        try:
            client = self._get_client()
            # Use correct API Gateway endpoint with camelCase fields
            api_run_id = getattr(self, '_api_run_id', run.id)
            response = client.patch(f"/api/test-runs/{api_run_id}", json={
                "endTime": run.end_time.isoformat() if run.end_time else None,
                "summary": {
                    "total": run.total,
                    "passed": run.passed,
                    "failed": run.failed,
                    "broken": run.broken,
                    "skipped": run.skipped,
                    "duration_ms": run.duration_ms,
                }
            })
            response.raise_for_status()
        except Exception as e:
            print(f"Warning: Failed to finalize run with API: {e}")
        finally:
            if self._client:
                self._client.close()
                self._client = None
    
    def report_test(self, test: TestResult) -> None:
        """Add test to batch and flush if needed."""
        if not self.config.api.enabled:
            return
        
        self._batch.append(test)
        
        if len(self._batch) >= self.config.api.batch_size:
            self._flush_batch()
    
    def _flush_batch(self) -> None:
        """Send batch of tests to API."""
        if not self._batch or not self._run:
            return
        
        try:
            client = self._get_client()
            api_run_id = getattr(self, '_api_run_id', self._run.id)
            # Send each test result individually
            for test in self._batch:
                response = client.post(
                    f"/api/test-runs/results",
                    json={
                        "runId": api_run_id,
                        "name": test.name,
                        "status": str(test.status),
                        "duration": test.duration_ms,
                        "metadata": test.labels or {},
                        "error": test.error_message,
                        "stackTrace": test.stack_trace,
                    }
                )
                response.raise_for_status()
        except Exception as e:
            print(f"Warning: Failed to send test results to API: {e}")
        finally:
            self._batch = []


class QAgenticReporter:
    """
    Main reporter that coordinates multiple output destinations.
    
    This is the primary interface for reporting test results.
    """
    
    _instance: Optional["QAgenticReporter"] = None
    _lock = threading.Lock()
    
    def __init__(self, config: Optional[QAgenticConfig] = None) -> None:
        self.config = config or get_config()
        self._reporters: List[BaseReporter] = []
        self._current_run: Optional[TestRunResult] = None
        
        # Initialize reporters based on config
        if self.config.features.console_output:
            self._reporters.append(ConsoleReporter(self.config))
        
        if self.config.local.enabled:
            self._reporters.append(JSONReporter(self.config))
            if "junit" in self.config.local.formats:
                self._reporters.append(JUnitReporter(self.config))
        
        if self.config.api.enabled:
            self._reporters.append(APIReporter(self.config))
    
    @classmethod
    def get_instance(cls, config: Optional[QAgenticConfig] = None) -> "QAgenticReporter":
        """Get singleton instance."""
        with cls._lock:
            if cls._instance is None:
                cls._instance = cls(config)
            return cls._instance
    
    @classmethod
    def reset(cls) -> None:
        """Reset singleton instance."""
        with cls._lock:
            cls._instance = None
    
    def start_run(
        self,
        name: Optional[str] = None,
        project_name: Optional[str] = None,
        environment: Optional[str] = None,
        **kwargs: Any,
    ) -> TestRunResult:
        """Start a new test run."""
        self._current_run = TestRunResult(
            name=name or f"run_{datetime.now(timezone.utc).strftime('%Y%m%d_%H%M%S')}",
            project_name=project_name or self.config.project_name,
            environment=environment or self.config.environment,
            start_time=datetime.now(timezone.utc),
            labels=self.config.labels.custom.copy(),
            **kwargs,
        )
        
        for reporter in self._reporters:
            reporter.start_run(self._current_run)
        
        return self._current_run
    
    def end_run(self) -> Optional[TestRunResult]:
        """End the current test run."""
        if not self._current_run:
            return None
        
        self._current_run.end_time = datetime.now(timezone.utc)
        self._current_run.duration_ms = (
            self._current_run.end_time - self._current_run.start_time
        ).total_seconds() * 1000
        
        for reporter in self._reporters:
            reporter.end_run(self._current_run)
        
        run = self._current_run
        self._current_run = None
        return run
    
    def report_test(self, test: TestResult) -> None:
        """Report a test result."""
        if self._current_run:
            self._current_run.add_test(test)
        
        for reporter in self._reporters:
            reporter.report_test(test)
    
    @property
    def current_run(self) -> Optional[TestRunResult]:
        """Get the current test run."""
        return self._current_run


def get_reporter(config: Optional[QAgenticConfig] = None) -> QAgenticReporter:
    """Get the global reporter instance."""
    return QAgenticReporter.get_instance(config)

File: packages/javascript/src/core/reporter.ts
/**
 * Reporter classes for outputting test results to various destinations.
 */

import * as fs from 'fs';
import * as path from 'path';
import axios from 'axios';
import { v4 as uuidv4 } from 'uuid';
import { getConfig, QAgenticConfig } from './config';
import { Status } from './status';
import type { TestResult, TestRunResult, StepResult } from './types';

/**
 * Base reporter interface.
 */
export interface BaseReporter {
  startRun(run: TestRunResult): void | Promise<void>;
  endRun(run: TestRunResult): void | Promise<void>;
  reportTest(test: TestResult): void | Promise<void>;
}

/**
 * Console reporter with colored output.
 */
export class ConsoleReporter implements BaseReporter {
  private config: QAgenticConfig;

  constructor(config?: QAgenticConfig) {
    this.config = config || getConfig();
  }

  startRun(run: TestRunResult): void {
    console.log('\n' + '='.repeat(60));
    console.log(`🚀 QAagentic Test Run - ${run.projectName}`);
    console.log(`Environment: ${run.environment}`);
    console.log('='.repeat(60) + '\n');
  }

  endRun(run: TestRunResult): void {
    const icon = run.failed === 0 ? '✅' : '❌';
    console.log('\n' + '='.repeat(60));
    console.log(`${icon} Test Run Complete - ${run.passRate.toFixed(1)}% Pass Rate`);
    console.log(`Passed: ${run.passed} | Failed: ${run.failed} | Skipped: ${run.skipped}`);
    console.log('='.repeat(60) + '\n');
  }

  reportTest(test: TestResult): void {
    if (!this.config.features.consoleOutput) return;

    const symbols: Record<Status, string> = {
      [Status.PASSED]: '✓',
      [Status.FAILED]: '✗',
      [Status.BROKEN]: '!',
      [Status.SKIPPED]: '○',
      [Status.PENDING]: '…',
      [Status.RUNNING]: '→',
      [Status.UNKNOWN]: '?',
    };

    const symbol = symbols[test.status as Status] || '?';
    console.log(`  ${symbol} ${test.name}`);

    if (test.errorMessage) {
      console.log(`    Error: ${test.errorMessage.slice(0, 100)}...`);
    }
  }
}

/**
 * JSON file reporter.
 */
export class JSONReporter implements BaseReporter {
  private config: QAgenticConfig;
  private outputDir: string;

  constructor(config?: QAgenticConfig) {
    this.config = config || getConfig();
    this.outputDir = this.config.local.outputDir;
  }

  startRun(run: TestRunResult): void {
    // Create output directory
    if (!fs.existsSync(this.outputDir)) {
      fs.mkdirSync(this.outputDir, { recursive: true });
    }

    // Clean previous results if configured
    if (this.config.local.cleanOnStart) {
      const files = fs.readdirSync(this.outputDir);
      for (const file of files) {
        if (file.endsWith('.json')) {
          fs.unlinkSync(path.join(this.outputDir, file));
        }
      }
    }
  }

  endRun(run: TestRunResult): void {
    if (!this.config.local.enabled || !this.config.local.formats.includes('json')) {
      return;
    }

    // Write run summary
    const runFile = path.join(this.outputDir, 'run.json');
    fs.writeFileSync(runFile, JSON.stringify(this.serializeRun(run), null, 2));

    // Write individual test results
    const testsDir = path.join(this.outputDir, 'tests');
    if (!fs.existsSync(testsDir)) {
      fs.mkdirSync(testsDir, { recursive: true });
    }

    for (const test of run.tests) {
      const testFile = path.join(testsDir, `${test.id}.json`);
      fs.writeFileSync(testFile, JSON.stringify(this.serializeTest(test), null, 2));
    }
  }

  reportTest(_test: TestResult): void {
    // Tests are written at end
  }

  private serializeRun(run: TestRunResult): Record<string, unknown> {
    return {
      ...run,
      startTime: run.startTime?.toISOString(),
      endTime: run.endTime?.toISOString(),
      tests: run.tests.map((t) => this.serializeTest(t)),
    };
  }

  private serializeTest(test: TestResult): Record<string, unknown> {
    return {
      ...test,
      startTime: test.startTime?.toISOString(),
      endTime: test.endTime?.toISOString(),
      steps: test.steps.map((s) => this.serializeStep(s)),
    };
  }

  private serializeStep(step: StepResult): Record<string, unknown> {
    return {
      ...step,
      startTime: step.startTime?.toISOString(),
      endTime: step.endTime?.toISOString(),
      children: step.children.map((c) => this.serializeStep(c)),
    };
  }
}

/**
 * JUnit XML reporter for CI/CD compatibility.
 */
export class JUnitReporter implements BaseReporter {
  private config: QAgenticConfig;
  private outputDir: string;

  constructor(config?: QAgenticConfig) {
    this.config = config || getConfig();
    this.outputDir = this.config.local.outputDir;
  }

  startRun(_run: TestRunResult): void {
    if (!fs.existsSync(this.outputDir)) {
      fs.mkdirSync(this.outputDir, { recursive: true });
    }
  }

  endRun(run: TestRunResult): void {
    if (!this.config.local.enabled || !this.config.local.formats.includes('junit')) {
      return;
    }

    const xml = this.generateXml(run);
    const junitFile = path.join(this.outputDir, 'junit.xml');
    fs.writeFileSync(junitFile, xml);
  }

  reportTest(_test: TestResult): void {
    // Tests are written at end
  }

  private generateXml(run: TestRunResult): string {
    const escape = (str: string) =>
      str
        .replace(/&/g, '&amp;')
        .replace(/</g, '&lt;')
        .replace(/>/g, '&gt;')
        .replace(/"/g, '&quot;');

    let xml = '<?xml version="1.0" encoding="UTF-8"?>\n';
    xml += `<testsuite name="${escape(run.projectName)}" `;
    xml += `tests="${run.total}" `;
    xml += `failures="${run.failed}" `;
    xml += `errors="${run.broken}" `;
    xml += `skipped="${run.skipped}" `;
    xml += `time="${(run.durationMs / 1000).toFixed(3)}" `;
    xml += `timestamp="${run.startTime?.toISOString() || ''}">\n`;

    for (const test of run.tests) {
      xml += `  <testcase name="${escape(test.name)}" `;
      xml += `classname="${escape(test.fullName)}" `;
      xml += `time="${(test.durationMs / 1000).toFixed(3)}"`;

      if (test.status === Status.PASSED) {
        xml += '/>\n';
      } else if (test.status === Status.FAILED) {
        xml += '>\n';
        xml += `    <failure message="${escape(test.errorMessage || 'Test failed')}" `;
        xml += `type="${escape(test.errorType || 'AssertionError')}">`;
        if (test.stackTrace) {
          xml += escape(test.stackTrace);
        }
        xml += '</failure>\n';
        xml += '  </testcase>\n';
      } else if (test.status === Status.BROKEN) {
        xml += '>\n';
        xml += `    <error message="${escape(test.errorMessage || 'Test error')}" `;
        xml += `type="${escape(test.errorType || 'Error')}">`;
        if (test.stackTrace) {
          xml += escape(test.stackTrace);
        }
        xml += '</error>\n';
        xml += '  </testcase>\n';
      } else if (test.status === Status.SKIPPED) {
        xml += '>\n';
        xml += `    <skipped${test.errorMessage ? ` message="${escape(test.errorMessage)}"` : ''}/>\n`;
        xml += '  </testcase>\n';
      } else {
        xml += '/>\n';
      }
    }

    xml += '</testsuite>\n';
    return xml;
  }
}

/**
 * API reporter that sends results to QAagentic server.
 */
export class APIReporter implements BaseReporter {
  private config: QAgenticConfig;
  private batch: TestResult[] = [];
  private currentRun?: TestRunResult;

  constructor(config?: QAgenticConfig) {
    this.config = config || getConfig();
  }

  async startRun(run: TestRunResult): Promise<void> {
    if (!this.config.api.enabled) {
      console.log('[APIReporter] API reporting disabled');
      return;
    }

    this.currentRun = run;
    this.batch = [];

    try {
      console.log('[APIReporter] Starting test run:', run.id);
      const url = `${this.config.api.url}/api/v1/runs`;
      console.log('[APIReporter] POST to:', url);
      
      const response = await axios.post(
        url,
        {
          id: run.id,
          name: run.name,
          project_name: run.projectName,
          environment: run.environment,
          start_time: run.startTime?.toISOString(),
          labels: run.labels,
          ci_build_id: run.ciBuildId,
          branch: run.branch,
          commit_hash: run.commitHash,
        },
        {
          headers: {
            'Content-Type': 'application/json',
            'X-API-Key': this.config.api.key || '',
            'X-Project': this.config.projectName,
          },
          timeout: this.config.api.timeout,
        }
      );
      console.log('[APIReporter] Test run started successfully:', response.status);
    } catch (error: any) {
      console.warn('[APIReporter] Failed to register run with API:');
      console.warn('  Error:', error.message);
      console.warn('  Status:', error.response?.status);
      console.warn('  Data:', error.response?.data);
    }
  }

  async endRun(run: TestRunResult): Promise<void> {
    if (!this.config.api.enabled) return;

    try {
      console.log('[APIReporter] endRun called, batch size:', this.batch.length);
      
      // Flush remaining batch - ALWAYS flush, even if batch is not full
      if (this.batch.length > 0) {
        console.log('[APIReporter] Flushing remaining', this.batch.length, 'test results');
        await this.flushBatch();
        console.log('[APIReporter] Batch flushed, waiting for processing...');
        await new Promise((resolve) => setTimeout(resolve, 1500));
      }

      console.log('[APIReporter] Finalizing run with test counts:', {
        total: run.total,
        passed: run.passed,
        failed: run.failed,
        broken: run.broken,
        skipped: run.skipped,
      });
      
      const patchData = {
        end_time: run.endTime?.toISOString(),
        duration_ms: run.durationMs,
        total: run.total,
        passed: run.passed,
        failed: run.failed,
        broken: run.broken,
        skipped: run.skipped,
        status: 'completed',
      };
      
      console.log('[APIReporter] PATCH to:', `${this.config.api.url}/api/v1/runs/${run.id}`);
      console.log('[APIReporter] PATCH data:', patchData);
      
      const response = await axios.patch(
        `${this.config.api.url}/api/v1/runs/${run.id}`,
        patchData,
        {
          headers: {
            'Content-Type': 'application/json',
            'X-API-Key': this.config.api.key || '',
            'X-Project': this.config.projectName,
          },
          timeout: this.config.api.timeout,
        }
      );
      console.log('[APIReporter] Run finalized successfully:', response.status);
    } catch (error: any) {
      console.warn('[APIReporter] Failed to finalize run with API:');
      console.warn('  Error:', error.message);
      console.warn('  Status:', error.response?.status);
      console.warn('  Data:', error.response?.data);
    }
  }

  async reportTest(test: TestResult): Promise<void> {
    console.log('[APIReporter] reportTest called for:', test.name, 'API enabled:', this.config.api.enabled);
    if (!this.config.api.enabled) {
      console.log('[APIReporter] API reporting disabled, skipping test report');
      return;
    }

    console.log('[APIReporter] Adding test to batch:', test.name, '- batch size:', this.batch.length + 1);
    this.batch.push(test);

    if (this.batch.length >= this.config.api.batchSize) {
      console.log('[APIReporter] Batch size reached, flushing batch');
      await this.flushBatch();
    } else {
      console.log('[APIReporter] Batch size:', this.batch.length, '/', this.config.api.batchSize);
    }
  }

  private async flushBatch(): Promise<void> {
    if (this.batch.length === 0 || !this.currentRun) return;

    try {
      console.log('[APIReporter] Flushing batch of', this.batch.length, 'test results');
      const response = await axios.post(
        `${this.config.api.url}/api/v1/runs/${this.currentRun.id}/results`,
        this.batch.map((t) => ({
          id: t.id,
          name: t.name,
          full_name: t.fullName,
          status: t.status,
          duration_ms: t.durationMs,
          start_time: t.startTime?.toISOString(),
          end_time: t.endTime?.toISOString(),
          file_path: t.filePath,
          error_message: t.errorMessage,
          stack_trace: t.stackTrace,
          error_type: t.errorType,
          attachments: t.attachments,
        })),
        {
          headers: {
            'Content-Type': 'application/json',
            'X-API-Key': this.config.api.key || '',
          },
          timeout: this.config.api.timeout,
        }
      );
      console.log('[APIReporter] Batch flushed successfully:', response.status);
    } catch (error: any) {
      console.warn('[APIReporter] Failed to send test results to API:');
      console.warn('  Error:', error.message);
      console.warn('  Status:', error.response?.status);
      console.warn('  Data:', error.response?.data);
    } finally {
      this.batch = [];
    }
  }
}

/**
 * Main reporter that coordinates multiple output destinations.
 */
export class QAgenticReporter {
  private static instance: QAgenticReporter | null = null;
  private config: QAgenticConfig;
  private reporters: BaseReporter[] = [];
  private currentRun: TestRunResult | null = null;

  constructor(config?: QAgenticConfig) {
    this.config = config || getConfig();

    console.log('[QAgenticReporter] Initializing with config:', {
      projectName: this.config.projectName,
      environment: this.config.environment,
      apiEnabled: this.config.api.enabled,
      apiUrl: this.config.api.url,
    });

    // Initialize reporters based on config
    if (this.config.features.consoleOutput) {
      this.reporters.push(new ConsoleReporter(this.config));
      console.log('[QAgenticReporter] Added ConsoleReporter');
    }

    if (this.config.local.enabled) {
      this.reporters.push(new JSONReporter(this.config));
      console.log('[QAgenticReporter] Added JSONReporter');
      if (this.config.local.formats.includes('junit')) {
        this.reporters.push(new JUnitReporter(this.config));
        console.log('[QAgenticReporter] Added JUnitReporter');
      }
    }

    if (this.config.api.enabled) {
      this.reporters.push(new APIReporter(this.config));
      console.log('[QAgenticReporter] Added APIReporter');
    }

    console.log('[QAgenticReporter] Initialization complete with', this.reporters.length, 'reporters');
  }

  /**
   * Get singleton instance.
   */
  static getInstance(config?: QAgenticConfig): QAgenticReporter {
    if (!QAgenticReporter.instance) {
      QAgenticReporter.instance = new QAgenticReporter(config);
    }
    return QAgenticReporter.instance;
  }

  /**
   * Reset singleton instance.
   */
  static reset(): void {
    QAgenticReporter.instance = null;
  }

  /**
   * Start a new test run.
   */
  async startRun(options: Partial<TestRunResult> = {}): Promise<TestRunResult> {
    this.currentRun = {
      id: options.id || uuidv4(),
      name: options.name || `run_${new Date().toISOString().replace(/[:.]/g, '')}`,
      projectName: options.projectName || this.config.projectName,
      environment: options.environment || this.config.environment,
      startTime: new Date(),
      durationMs: 0,
      tests: [],
      total: 0,
      passed: 0,
      failed: 0,
      broken: 0,
      skipped: 0,
      passRate: 0,
      labels: { ...this.config.labels.custom, ...options.labels },
      parameters: options.parameters || {},
      ciBuildId: options.ciBuildId,
      ciBuildUrl: options.ciBuildUrl,
      branch: options.branch,
      commitHash: options.commitHash,
    };

    for (const reporter of this.reporters) {
      await reporter.startRun(this.currentRun);
    }

    return this.currentRun;
  }

  /**
   * End the current test run.
   */
  async endRun(): Promise<TestRunResult | null> {
    if (!this.currentRun) return null;

    this.currentRun.endTime = new Date();
    this.currentRun.durationMs =
      this.currentRun.endTime.getTime() - (this.currentRun.startTime?.getTime() || 0);
    this.currentRun.passRate =
      this.currentRun.total > 0 ? (this.currentRun.passed / this.currentRun.total) * 100 : 0;

    for (const reporter of this.reporters) {
      await reporter.endRun(this.currentRun);
    }

    const run = this.currentRun;
    this.currentRun = null;
    return run;
  }

  /**
   * Report a test result.
   */
  async reportTest(test: TestResult): Promise<void> {
    if (!this.currentRun) {
      console.warn('[QAgenticReporter] No active run to report test');
      return;
    }

    this.currentRun.tests.push(test);
    this.currentRun.total++;

    switch (test.status) {
      case 'passed':
        this.currentRun.passed++;
        break;
      case 'failed':
        this.currentRun.failed++;
        break;
      case 'broken':
        this.currentRun.broken++;
        break;
      case 'skipped':
        this.currentRun.skipped++;
        break;
    }

    console.log('[QAgenticReporter] Test counts:', {
      total: this.currentRun.total,
      passed: this.currentRun.passed,
      failed: this.currentRun.failed,
      broken: this.currentRun.broken,
      skipped: this.currentRun.skipped,
    });

    for (const reporter of this.reporters) {
      await reporter.reportTest(test);
    }
  }

  /**
   * Get the current test run.
   */
  getCurrentRun(): TestRunResult | null {
    return this.currentRun;
  }
}

/**
 * Get the global reporter instance.
 */
export function getReporter(config?: QAgenticConfig): QAgenticReporter {
  return QAgenticReporter.getInstance(config);
}
